strict internal libc compat

fbrausse opened the pull request:
This PR enforces usage of headers of ESBMC's internal libc for `c2goto` as well as `esbmc` when parsing source files (with the clang-c-frontend). If this approach turns out to be successful (no clashes with system headers on any supported OS),
- we can remove unchanged copies of headers such as bits/pthreadtypes.h, see #600, and
- we will get compiler errors for incompatibilities between the OS headers and ESBMC's; thereby
- discrepancies between any pre-built operational models and the translated user sources should be minimized, that is, for any library we (want to) model (e.g., POSIX threads), the sources will be translated with types corresponding to ESBMC's libc.

~It is work-in-progress at the moment.~

fbrausse committed:
add --sysroot parameter to esbmc and c2goto; pass it through to clang

fbrausse committed:
[c2goto] typo in comment

fbrausse committed:
[c2goto] add access to header-dir of internal libc

fbrausse committed:
[clang-c-frontend] pass internal libc header-dir to clang

fbrausse committed:
[c2goto] library: rename *.{hs => h}, add __clang__ version of stddef.h, move pthreadtypes.h to bits/

fbrausse committed:
[c2goto] ignore headers/ for clang-format

fbrausse committed:
[c2goto] include libc headers in c2goto -> default to internal libc

fbrausse committed:
add --dirafter PATH option to esbmc, c2goto, config and clang-c-frontend

fbrausse commented:
Linux:
>  The following tests FAILED:
-	263 - regression-esbmc-unix2-11_olstate (Failed)
  Broken test-case in my opinion: can't #define uint8_t and uint16_t to int and expect standard headers to agree.
-	289 - regression-esbmc-unix2-23_picosat-846_01 (Failed)
	290 - regression-esbmc-unix2-23_picosat-846_02 (Failed)
	292 - regression-esbmc-unix2-23_picosat-846_04 (Failed)
	293 - regression-esbmc-unix2-23_picosat-846_05 (Failed)
  need to fix `union pthread_attr_t` in this PR.
-	1552 - regression-esbmc-cpp/cpp-cpp_with_pthread_01 (Failed)
	1553 - regression-esbmc-cpp/cpp-cpp_with_pthread_02 (Failed)
  `KNOWNBUG` -> `CORE`

*Edit*: done.

fbrausse committed:
[regression] C++ pthread only still fails with --no-library; added tests for that

fbrausse committed:
[c2goto lib] header: name 'union pthread_attr_t' in line with Glibc's prototypes in <bits/types/sigevent_t.h>

Fixes regression/esbmc-unix2/23_picosat-846_0[1-5] for glibc-2.33.

fbrausse committed:
[regression] make esbmc-unix2/11_olstate more standards compliant

fbrausse commented:
Obviously I'll clean up the history to remove my failed attempts. Many of those have been to figure out what is really meant by setting include paths (e.g. for the macOS build). The following convention is introduced in the clang-c-frontend instead of just using `-I`:
- ESBMC's libc path is the first `-isystem` parameter to clang
- Clang's header path is the second `-isystem` parameter to clang
- the macOS build now uses `c2goto --sysroot ${C2GOTO_INCLUDE_DIR}/../..` -- this is obviously in need of cleanup, ie, rename the CMake parameter; note that `--sysroot` replaces the directories in step 5 below

We move to `-isystem` instead of `-I` in order to bring ESBMC's behaviour regarding include-paths closer to the order in which clang/gcc handle #include-s, thereby freeing `-I` for users alone:
1. `"..."` is looked up in the directory of the source
2. `"..."` is looked up in the `-iquote` list from left to right
3. both forms `"..."` and `<...>` are looked up in the `-I` list from left to right
4. both forms are looked up in the `-isystem` list from left to right
5. look up in standard system directories
6. look up in the `-idirafter` list from left to right

fbrausse commented:
PS: I should add that accepting this PR is a step towards supporting the different OS libc-s more whole-heartedly - or replacing them by an own one, see also #183.

lucasccordeiro commented:
@fbrausse: Is this PR ready to review?

fbrausse committed:
[cmake] switch C2GOTO_{INCLUDE_DIR => SYSROOT} from -I to --sysroot

The flag -I makes system headers take precedence over esbmc's headers of its
operational models, while --sysroot puts esbmc's headers first.

fbrausse committed:
[cmake] rm special case when CMAKE_LIBRARY_ARCHITECTURE is set

This construction was supposed to produce a path like

	/usr/include/x86_64-linux-gnu

on multilib (also known as multiarch) systems in order to include headers
corresponding to the multilib "architecture" for which stuff is being built.

It has a few problems:

a) This setting is only meaningful for library search paths and was not
   intended to be used in such a way, cf. CMake docs.
b) The compiler is already doing "the right thing" on multilib systems
   depending on the particular target configuration (i.e. -m32 vs. -m64).
c) Enabling it unconditionally (see b) above) may lead to inclusion of a
   64-bit header in -m32 mode.

That's why we drop it.

fbrausse commented:
Yes.

lucasccordeiro reviewed (approved):
LGTM.

lucasccordeiro commented:
@fbrausse: The following test cases are failing in MacOS. Is it a bug or a feature?

The following tests FAILED:
	 27 - regression-esbmc-unix-00_account_01 (Failed)
	 28 - regression-esbmc-unix-00_account_02 (Failed)
	 34 - regression-esbmc-unix-00_atomicity06 (Failed)
	 36 - regression-esbmc-unix-00_bbuf_01 (Failed)
	 37 - regression-esbmc-unix-00_bbuf_02 (Failed)
	 38 - regression-esbmc-unix-00_bbuf_02_new (Failed)
	 45 - regression-esbmc-unix-00_race07 (Failed)
	 52 - regression-esbmc-unix-00_race14 (Failed)
	 55 - regression-esbmc-unix-00_race17 (Failed)
	 56 - regression-esbmc-unix-00_race18 (Failed)
	 57 - regression-esbmc-unix-00_race19 (Failed)
	 58 - regression-esbmc-unix-00_race20 (Failed)
	 73 - regression-esbmc-unix-01_malloc_02 (Failed)
	 76 - regression-esbmc-unix-01_malloc_05 (Failed)
	 77 - regression-esbmc-unix-01_malloc_06 (Failed)
	 81 - regression-esbmc-unix-01_malloc_10 (Failed)
	 87 - regression-esbmc-unix-01_malloc_16 (Failed)
	125 - regression-esbmc-unix-03_boundedBuffer (Failed)
	126 - regression-esbmc-unix-03_bzip2smp_comb_01 (Failed)
	127 - regression-esbmc-unix-03_bzip2smp_comb_01_new (Failed)
	130 - regression-esbmc-unix-03_falcon (Failed)
	140 - regression-esbmc-unix-03_twostage_01 (Failed)
	141 - regression-esbmc-unix-03_twostage_02 (Failed)
	145 - regression-esbmc-unix-03_wronglock_01 (Failed)
	146 - regression-esbmc-unix-03_wronglock_02 (Failed)
	147 - regression-esbmc-unix-03_wronglock_03 (Failed)
	149 - regression-esbmc-unix-04_bzip2smp_03 (Failed)
	165 - regression-esbmc-unix-github_397_3 (Failed)
	166 - regression-esbmc-unix-github_397_4 (Failed)
	176 - regression-esbmc-unix2-01_pthread11 (Failed)
	177 - regression-esbmc-unix2-01_pthread12 (Failed)
	205 - regression-esbmc-unix2-01_pthread38 (Failed)
	216 - regression-esbmc-unix2-01_pthread48 (Failed)
	219 - regression-esbmc-unix2-01_pthread50 (Failed)
	220 - regression-esbmc-unix2-01_pthread51 (Failed)
	221 - regression-esbmc-unix2-01_pthread52 (Failed)
	222 - regression-esbmc-unix2-01_pthread53 (Failed)
	223 - regression-esbmc-unix2-01_pthread54 (Failed)
	234 - regression-esbmc-unix2-01_pthread64 (Failed)
	235 - regression-esbmc-unix2-01_pthread65 (Failed)
	237 - regression-esbmc-unix2-01_pthread67 (Failed)
	260 - regression-esbmc-unix2-11_lu-fig2.fixed (Failed)
	305 - regression-esbmc-unix2-github_213 (Failed)
Errors while running CTest
Error: Process completed with exit code 8.

fbrausse commented:
> The following test cases are failing in MacOS. Is it a bug or a feature?

The logs show pretty much the discrepencies between ESBMC's headers and (in this case, since our headers come from glibc) the OS X headers. `union pthread_attr_t` vs. `struct _opaque_pthread_attr_t`, etc. These show explicitely where assumptions before had silently worked - just because the space occupied in memory has not been overrun. Assumptions were:
- internal operational models are defined based on the glibc-provided pthread types defined in `c2goto/headers`.
- user code (regression tests in this case) have been translated using system definitions - which differ between libc-s.

I don't expect this PR to be merged quickly, I suspect it is going to take time to sort this out - in particular since I don't have an OS X machine available.

Essentially, I see 2 ways to fix these issues:
1. get rid of most libc-specific headers (like `bits/pthreadtypes.h`, but not compiler-provided ones like `stddef.h`) and model the behaviour for every supported libc/OS separately (e.g., via `#ifdef __APPLE__`, `#ifdef __GLIBC__`, etc. *edit* Free-/Cheribsd to come)
2. settle for one specific implementation (does not have to be an existing one, just one that follows all assumptions made by standards like POSIX) and provide headers defining the concrete types used in the (single) operational model.

fbrausse commented:
The choice between 1. and 2. (which I as a newcomer cannot make) affects whether ESBMC is to support code that (preprocessor-)branches on different libc-s or we only want to support code that is oblivious to the concrete library implementation (i.e., it doesn't care how pthread_attr_t is defined, as long as the API works) and thus accepts ESBMC's models independent of the system.

Obviously when there is no standard - like e.g, OpenSSL 1.1.x vs. 3.y vs. LibreSSL vs..., choice 1. is better for users while putting a burden on ESBMC's maintainers. On the other hand standards exist for a reason - and following a moving target like the *SSL implementations could be considered a user's choice when employing a general C/C++ model checker.

kunjsong01 commented on src/clang-c-frontend/clang_c_language.cpp:
```diff
@@ -43,7 +43,15 @@ void clang_c_languaget::build_compiler_args(const std::string &tmp_dir)
+    compiler_args.push_back(*libc_headers);
+  }
+
+  compiler_args.push_back("-isystem");
```
Same as line 49. pehaps make it common?

fbrausse replied:
We would have the same `if` plus another std::vector. I don't think it's worth it at this point. We don't even expose `-isystem` as a cmdline parameter at this point.

fbrausse committed:
[cmdline] also parse options from envvar ESBMC_OPTS when set

This allows to easily pass -I, -D or --sysroot options to esbmc if it is run
as part of another tool like ctest, testing_tool.py or the esbmc_wrapper.py.

The content of the environment variable ESBMC_OPTS is split at white-space
into a sequence of strings according to a simple interpretation of shell rules
taking ",',\ into account. These tokens are then parsed like normal command
line parameters before the actual ones.

No checking of validity of options or exclusive settings (like a particular
solver or endianness) is performed. This means conflicts could arise in
particular for regression tests where the options specified by test-cases
are more or less arbitrary while the envvar setting remains the same.

There are some ways to remedy this deficit. E.g., by switching to a more
flexible way of parsing arguments (i.e. later exclusive options overwrite
former ones) or by explicitely specifying exclusivity and precedence. This
patch does nothing of the sort.

fbrausse commented:
@lucasccordeiro explained to me in private conversation:
> the operational models are an abstract representation of the original libraries
So, as long as we capture the original semantics of the library, we should be okay

which I interpret as going for ~option 1~ *edit:* sorry, misread: option 2 above. Incidentally, the regressions failing on CI for #615 profit from this as well.

fbrausse committed:
[elibc] pthreads: simplify definitions for operational model

This removes all platform-specific definitions/declarations from the internal
model of pthreads. The goal is to unify the interpretation of user code across
all platforms/system. This patch therefore removes the Glibc-specific parts
as well as unnecessary fields from the defined types, as well as their
embedding into a union.

The second change concerns the global variable 'head' defined in pthread_lib.c:
It was exported as an external symbol before, which the regression test
nonz3/01_cbmc_Malloc18 did uncover, as there is a conflict with the symbol of
the same name declared there. Therefore, pthread_lib's 'head' now has internal
linkage.

fbrausse committed:
[elibc] add 2 glibc-specific headers in bits/ redirecting to <pthread.h>

fbrausse committed:
[elibc] add 2 OS X-specific headers in sys/ redirecting to <pthread.h>

sys/_types/_pthread_attr_t.h is for versions 10.x, while
sys/_pthread/_pthread_attr_t.h is for 11.x

fbrausse committed:
clang-format

fbrausse committed:
[elibc] add more OS X-specific headers redirecting to <pthread.h>

fbrausse committed:
temporarily let c2goto message on Debug to fix build on OS X

fbrausse committed:
[elibc] fix: add missing OS X headers to cmake

fbrausse committed:
[elibc] fix: add remaining OS X-specific pthread headers

fbrausse committed:
[elibc] do not refer to dummy headers, use <pthread.h> directly

fbrausse committed:
[config] recognize ARM host archs

fbrausse committed:
[goto-symex] exec-state: do not use "int $3" on 32-bit ARM

fbrausse committed:
[elibc] pthread.h: always expose declaration of pthread_mutexattr_settype() regardless of feature test macros

fbrausse committed:
[elibc] export PTHREAD_MUTEX_(NORMAL|RECURSIVE|ERRORCHECK|DEFAULT) independent of feature test macros

Background: the operational model behaves as a new libc implementation towards
user code. As such there is no need for it to respect feature requests specific
to, e.g., Glibc.

fbrausse commented:
The macOS issues are fixed now by simplifying its types, and removing all architecture-specific #ifdefs.

fbrausse committed:
Revert "temporarily let c2goto message on Debug to fix build on OS X"

This reverts commit e9f177c4e9408046cd7284585f6ee4b0c35a642b.

lucasccordeiro reviewed (approved):
@fbrausse: This is a very nice piece of work! Well done!

I believe this PR will improve the performance of our concurrency checking quite a lot!

Can I ask you to run this PR over the concurrency category only? 

You can use these parameters:

1) 
  if concurrency:
    command_line += " --no-por --context-bound 2  "

2)
  if concurrency:
    command_line += " --no-por --context-bound 2  --no-goto-merge "

This PR contains the latest results for comparison: https://github.com/esbmc/esbmc/pull/617

lucasccordeiro commented:
@fbrausse: could you also please rebase this branch?

lucasccordeiro commented:
This PR with  `--incremental-bmc --no-por --context-bound 2 --no-goto-merge`:

```
Statistics:            763 Files
  correct:             336
    correct true:       69
    correct false:     267
  incorrect:             0
    incorrect true:      0
    incorrect false:     0
  unknown:             427
  Score:               405 (max: 1160)
```

Master with `--incremental-bmc --no-por --context-bound 2 --no-goto-merge`:

```
Statistics:            763 Files
  correct:             336
    correct true:       69
    correct false:     267
  incorrect:             0
    incorrect true:      0
    incorrect false:     0
  unknown:             427
  Score:               405 (max: 1160)
```

lucasccordeiro commented:
This PR with  `--incremental-bmc --no-por --context-bound 2 --no-goto-merge`:

```
Statistics:            763 Files
  correct:             336
    correct true:       69
    correct false:     267
  incorrect:             0
    incorrect true:      0
    incorrect false:     0
  unknown:             427
  Score:               405 (max: 1160)
```

Master with `--incremental-bmc --no-por --context-bound 2 --no-goto-merge`:

```
Statistics:            763 Files
  correct:             336
    correct true:       69
    correct false:     267
  incorrect:             0
    incorrect true:      0
    incorrect false:     0
  unknown:             427
  Score:               405 (max: 1160)
```
